server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/DeletedMessagesVaultRoutes.java

```diff
@@ -220,21 +157,39 @@ public void define(Service service) {
         @ApiResponse(code = HttpStatus.NOT_FOUND_404, message = "Not found - requested user does not exist"),
         @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side.")
     })
-    private TaskIdDto userActions(Request request, Response response) throws JsonExtractException {
-        VaultAction requestedAction = extractParam(request, ACTION_QUERY_PARAM, this::getVaultAction);
-        validateAction(requestedAction, VaultAction.RESTORE, VaultAction.EXPORT);
-
-        Task requestedTask = generateUserTask(requestedAction, request);
-        TaskId taskId = taskManager.submit(requestedTask);
-        return TaskIdDto.respond(response, taskId);
+    private Route userActions() {
+        TaskGenerator export = TaskGenerator.builder()
+            .registrationKey(EXPORT_REGISTRATION_KEY)
+            .task(request -> {
+                Username username = extractUser(request);
+                validateUserExist(username);
+                Query query = translate(jsonExtractor.parse(request.body()));
+                return new DeletedMessagesVaultExportTask(vaultExport, username, query, extractMailAddress(request));
+            })
+            .build();
+        TaskGenerator restore = TaskGenerator.builder()
+            .registrationKey(RESTORE_REGISTRATION_KEY)
+            .task(request -> {
+                Username username = extractUser(request);
+                validateUserExist(username);
+                Query query = translate(jsonExtractor.parse(request.body()));
+                return new DeletedMessagesVaultRestoreTask(vaultRestore, username, query);
+            })
```

Comment:
these tasks would benefit a method extraction I guess.

Reply:
Probably (it's a poc)